File: src/sim/shading.ts
import type { GridSpec, Polygon2D, Ring2D, UnitVec3, Vec2M } from '../types/geo'
import type { Occluder } from '../types/pv'
import type { Degrees, Fraction, Meters, Radians } from '../types/units'
import { cosDeg, sinDeg } from './math'

export const projectPanelToGround: (panel: Occluder, sun: UnitVec3) => Polygon2D = (panel, sun) => {
  if (sun.z <= 1e-6) return { exterior: [], holes: [] }
  const exterior: Vec2M[] = panel.corners.vertices.map((corner) => {
    const t = corner.zM / sun.z
    return { xM: (corner.xM - sun.x * t) as Meters, yM: (corner.yM - sun.y * t) as Meters }
  })
  return { exterior, holes: [] }
}

const ringContains = (ring: Ring2D, point: Vec2M): boolean => {
  let inside = false
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i, i += 1) {
    const pi = ring[i] as Vec2M
    const pj = ring[j] as Vec2M
    const crosses =
      pi.yM > point.yM !== pj.yM > point.yM &&
      point.xM < ((pj.xM - pi.xM) * (point.yM - pi.yM)) / (pj.yM - pi.yM) + pi.xM
    if (crosses) inside = !inside
  }
  return inside
}

export const pointInPolygon: (point: Vec2M, polygon: Polygon2D) => boolean = (point, polygon) =>
  ringContains(polygon.exterior, point) && !polygon.holes.some((hole) => ringContains(hole, point))

interface GroundShadow {
  readonly polygon: Polygon2D
  readonly minX: number
  readonly maxX: number
  readonly minY: number
  readonly maxY: number
  readonly transmittance: Fraction
}

const ringBounds = (ring: Ring2D): Omit<GroundShadow, 'polygon' | 'transmittance'> => {
  let minX = Infinity
  let maxX = -Infinity
  let minY = Infinity
  let maxY = -Infinity
  for (const p of ring) {
    minX = Math.min(minX, p.xM)
    maxX = Math.max(maxX, p.xM)
    minY = Math.min(minY, p.yM)
    maxY = Math.max(maxY, p.yM)
  }
  return { minX, maxX, minY, maxY }
}

export const beamVisibilityRaster: (
  grid: GridSpec,
  panels: readonly Occluder[],
  sun: UnitVec3,
  moduleTransmittance: Fraction,
  subSamplesPerCell: number,
) => Float32Array = (grid, panels, sun, moduleTransmittance, subSamplesPerCell) => {
  const result = new Float32Array(grid.cols * grid.rows)
  if (sun.z <= 1e-6) return result

  const shadows: GroundShadow[] = []
  for (const panel of panels) {
    const polygon = projectPanelToGround(panel, sun)
    if (polygon.exterior.length === 0) continue
    shadows.push({
      polygon,
      ...ringBounds(polygon.exterior),
      transmittance: panel.transmittance ?? moduleTransmittance,
    })
  }

  const subSamples = Math.max(1, subSamplesPerCell)
  const totalSamples = subSamples * subSamples
  const step = grid.cellSizeM / subSamples

  for (let row = 0; row < grid.rows; row += 1) {
    const cellMinY = grid.extent.minYM + row * grid.cellSizeM
    for (let col = 0; col < grid.cols; col += 1) {
      const cellMinX = grid.extent.minXM + col * grid.cellSizeM
      let total = 0
      for (let si = 0; si < subSamples; si += 1) {
        const sampleX = cellMinX + (si + 0.5) * step
        for (let sj = 0; sj < subSamples; sj += 1) {
          const sampleY = cellMinY + (sj + 0.5) * step
          const point: Vec2M = { xM: sampleX as Meters, yM: sampleY as Meters }
          // the open figure is 1; a sample under one or more shadows keeps the smallest of their
          // transmittances, so a ray through two faces of one crown counts once
          let sampleTransmittance = 1
          for (const shadow of shadows) {
            if (
              sampleX >= shadow.minX &&
              sampleX <= shadow.maxX &&
              sampleY >= shadow.minY &&
              sampleY <= shadow.maxY &&
              pointInPolygon(point, shadow.polygon)
            ) {
              if (shadow.transmittance < sampleTransmittance)
                sampleTransmittance = shadow.transmittance
              if (sampleTransmittance === 0) break
            }
          }
          total += sampleTransmittance
        }
      }
      result[row * grid.cols + col] = total / totalSamples
    }
  }
  return result
}

export const shadedGroundFractionInfiniteRows: (
  collectorWidthM: Meters,
  pitchM: Meters,
  tiltDeg: Degrees,
  profileAngleRad: Radians,
  side: -1 | 0 | 1,
) => Fraction = (collectorWidthM, pitchM, tiltDeg, profileAngleRad, side) => {
  if (profileAngleRad <= 0) return 1 as Fraction
  const shaded =
    (collectorWidthM / pitchM) *
    Math.abs(cosDeg(tiltDeg) + (side * sinDeg(tiltDeg)) / Math.tan(profileAngleRad))
  return Math.min(1, shaded) as Fraction
}

// the Sun's mean angular diameter, 0.533 deg limb to limb, so the width needs no doubling
const SUN_ANGULAR_DIAMETER_RAD = 0.0093

/** The penumbra's width across the ray at a slant distance; on the ground it is 1/sin(elevation) wider */
export const penumbraWidthM: (slantDistanceM: Meters) => Meters = (slantDistanceM) =>
  (SUN_ANGULAR_DIAMETER_RAD * slantDistanceM) as Meters

export const rowSelfShadeFraction: (
  collectorWidthM: Meters,
  pitchM: Meters,
  tiltDeg: Degrees,
  profileAngleRad: Radians,
  side: -1 | 0 | 1,
) => Fraction = (collectorWidthM, pitchM, tiltDeg, profileAngleRad, side) => {
  if (profileAngleRad <= 0) return 0 as Fraction
  const d =
    collectorWidthM *
    Math.abs(cosDeg(tiltDeg) + (side * sinDeg(tiltDeg)) / Math.tan(profileAngleRad))
  return d <= pitchM ? (0 as Fraction) : ((1 - pitchM / d) as Fraction)
}

File: crates/agv-sim/src/shading.rs
//! Where the panels put their shadows, ported from `src/sim/shading.ts`.
//!
//! This is the CPU reference for the visibility question the GPU kernel answers, and
//! `src/sim/gpu/webgl2.test.ts` already cross-checks the two against each other cell by cell. That
//! makes this the module whose semantics matter most: the shader is checked against it, so a
//! divergence here is a divergence in what the whole bake means.

use crate::geom::{GridSpec, Polygon2D, UnitVec3, Vec2M, Vec3M};
use crate::math::{cos_deg, sin_deg};

/// Cast one panel's corners onto z = 0 along the sun direction.
///
/// Returns an empty polygon when the sun is at or below the horizon, which the caller reads as
/// "nothing is shaded because nothing is lit". The 1e-6 floor is not a tolerance on the sun angle,
/// it is what stops the division below producing a shadow thousands of metres long.
pub fn project_panel_to_ground(corners: &[Vec3M], sun: UnitVec3) -> Polygon2D {
    if sun.z <= 1e-6 {
        return Polygon2D::default();
    }
    Polygon2D {
        exterior: corners
            .iter()
            .map(|corner| {
                let t = corner.z_m / sun.z;
                Vec2M {
                    x_m: corner.x_m - sun.x * t,
                    y_m: corner.y_m - sun.y * t,
                }
            })
            .collect(),
        holes: Vec::new(),
    }
}

/// Crossing-number test, which is winding-agnostic and is what the TypeScript uses.
fn ring_contains(ring: &[Vec2M], point: Vec2M) -> bool {
    let mut inside = false;
    let n = ring.len();
    if n == 0 {
        return false;
    }
    let mut j = n - 1;
    for i in 0..n {
        let pi = ring[i];
        let pj = ring[j];
        let crosses = ((pi.y_m > point.y_m) != (pj.y_m > point.y_m))
            && (point.x_m
                < ((pj.x_m - pi.x_m) * (point.y_m - pi.y_m)) / (pj.y_m - pi.y_m) + pi.x_m);
        if crosses {
            inside = !inside;
        }
        j = i;
    }
    inside
}

pub fn point_in_polygon(point: Vec2M, polygon: &Polygon2D) -> bool {
    ring_contains(&polygon.exterior, point)
        && !polygon.holes.iter().any(|hole| ring_contains(hole, point))
}

struct GroundShadow {
    polygon: Polygon2D,
    min_x: f64,
    max_x: f64,
    min_y: f64,
    max_y: f64,
    transmittance: f64,
}

fn ring_bounds(polygon: Polygon2D, transmittance: f64) -> GroundShadow {
    let mut min_x = f64::INFINITY;
    let mut max_x = f64::NEG_INFINITY;
    let mut min_y = f64::INFINITY;
    let mut max_y = f64::NEG_INFINITY;
    for p in &polygon.exterior {
        min_x = min_x.min(p.x_m);
        max_x = max_x.max(p.x_m);
        min_y = min_y.min(p.y_m);
        max_y = max_y.max(p.y_m);
    }
    GroundShadow {
        polygon,
        min_x,
        max_x,
        min_y,
        max_y,
        transmittance,
    }
}

/// Per cell, the fraction of the beam that reaches the ground.
///
/// Sub-sampled inside each cell, because a cell either side of a shadow edge is genuinely part
/// shaded and a single centre sample would quantise the edge to the cell size. A blocked sample
/// still passes a transmittance, which is what makes a semi-transparent quad a shade level,
/// not a switch: `transmittances[i]` for panel `i` where the slice reaches that far, else
/// `module_transmittance` for every panel, an empty slice being the common case. A sample under
/// more than one quad keeps the smallest of their transmittances, so a ray through two faces of
/// one crown counts once.
pub fn beam_visibility_raster(
    grid: &GridSpec,
    panels: &[Vec<Vec3M>],
    sun: UnitVec3,
    module_transmittance: f64,
    transmittances: &[f64],
    sub_samples_per_cell: usize,
) -> Vec<f32> {
    let mut result = vec![0.0f32; grid.cells()];
    if sun.z <= 1e-6 {
        return result;
    }

    let shadows: Vec<GroundShadow> = panels
        .iter()
        .enumerate()
        .filter_map(|(index, corners)| {
            let polygon = project_panel_to_ground(corners, sun);
            if polygon.exterior.is_empty() {
                return None;
            }
            let transmittance = transmittances
                .get(index)
                .copied()
                .unwrap_or(module_transmittance);
            Some(ring_bounds(polygon, transmittance))
        })
        .collect();

    let sub_samples = sub_samples_per_cell.max(1);
    let total_samples = sub_samples * sub_samples;
    let step = grid.cell_size_m / sub_samples as f64;

    for row in 0..grid.rows {
        let cell_min_y = grid.extent.min_y_m + row as f64 * grid.cell_size_m;
        for col in 0..grid.cols {
            let cell_min_x = grid.extent.min_x_m + col as f64 * grid.cell_size_m;
            let mut total = 0.0f64;
            for si in 0..sub_samples {
                let sample_x = cell_min_x + (si as f64 + 0.5) * step;
                for sj in 0..sub_samples {
                    let sample_y = cell_min_y + (sj as f64 + 0.5) * step;
                    let point = Vec2M {
                        x_m: sample_x,
                        y_m: sample_y,
                    };
                    // the open figure is 1; a sample under one or more shadows keeps the
                    // smallest of their transmittances, the bounding box first because it
                    // rejects almost every quad almost always
                    let mut sample_transmittance = 1.0f64;
                    for shadow in &shadows {
                        if sample_x >= shadow.min_x
                            && sample_x <= shadow.max_x
                            && sample_y >= shadow.min_y
                            && sample_y <= shadow.max_y
                            && point_in_polygon(point, &shadow.polygon)
                        {
                            if shadow.transmittance < sample_transmittance {
                                sample_transmittance = shadow.transmittance;
                            }
                            if sample_transmittance <= 0.0 {
                                break;
                            }
                        }
                    }
                    total += sample_transmittance;
                }
            }
            result[row * grid.cols + col] = (total / total_samples as f64) as f32;
        }
    }
    result
}

/// The closed form for infinite rows, which is what pins the raster above at its limits.
pub fn shaded_ground_fraction_infinite_rows(
    collector_width_m: f64,
    pitch_m: f64,
    tilt_deg: f64,
    profile_angle_rad: f64,
    side: f64,
) -> f64 {
    if profile_angle_rad <= 0.0 {
        return 1.0;
    }
    let shaded = (collector_width_m / pitch_m)
        * (cos_deg(tilt_deg) + (side * sin_deg(tilt_deg)) / profile_angle_rad.tan()).abs();
    shaded.min(1.0)
}

/// The sun is not a point. Half a degree of angular diameter is 0.0093 radians of half-angle, and
/// this is the width of the soft edge that produces at a given slant distance.
const SUN_ANGULAR_RADIUS_RAD: f64 = 0.0093;

pub fn penumbra_width_m(slant_distance_m: f64) -> f64 {
    SUN_ANGULAR_RADIUS_RAD * slant_distance_m
}

/// How much of a row its neighbour shades: a loss the array takes.
pub fn row_self_shade_fraction(
    collector_width_m: f64,
    pitch_m: f64,
    tilt_deg: f64,
    profile_angle_rad: f64,
    side: f64,
) -> f64 {
    if profile_angle_rad <= 0.0 {
        return 0.0;
    }
    let d = collector_width_m
        * (cos_deg(tilt_deg) + (side * sin_deg(tilt_deg)) / profile_angle_rad.tan()).abs();
    if d <= pitch_m {
        0.0
    } else {
        1.0 - pitch_m / d
    }
}
